File: cms-app-crawler-js/src/utils/SelectorRegistry.js
'use strict';

const fs = require('fs');
const path = require('path');

let _fallbackMap = new Map();
let _loaded = false;

const REGISTRY_PATH = path.join(__dirname, '../../resources/pom_registry.json');

function load(registryPath = REGISTRY_PATH) {
  if (_loaded) return;
  if (!fs.existsSync(registryPath)) {
    console.warn(`SelectorRegistry: registry not found at ${registryPath} — no fallbacks available`);
    _loaded = true;
    return;
  }
  const entries = JSON.parse(fs.readFileSync(registryPath, 'utf8'));
  _fallbackMap.clear();
  for (const entry of entries) {
    const key = entry.selectorKey;
    if (!key) continue;
    _fallbackMap.set(key, entry.selectorFallbacks || []);
  }
  _loaded = true;
  console.info(`SelectorRegistry loaded: ${_fallbackMap.size} entries from ${registryPath}`);
}

function getFallbacks(selectorKey) {
  if (!_loaded) load();
  return _fallbackMap.get(selectorKey) || [];
}

function reset() {
  _fallbackMap.clear();
  _loaded = false;
}

module.exports = { SelectorRegistry: { load, getFallbacks, reset } };

File: cms-app-crawler/src/main/java/utils/SelectorRegistry.java
package utils;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.logging.Logger;

/**
 * Loads pom_registry.json at test startup and provides fallback selector lookup
 * for self-healing in BasePage.findElement().
 *
 * Call SelectorRegistry.load(path) once in @BeforeClass, not per test.
 */
public class SelectorRegistry {

    private static final Logger log = Logger.getLogger(SelectorRegistry.class.getName());
    private static final Map<String, List<String>> fallbackMap = new HashMap<>();
    private static boolean loaded = false;

    public static void load(String registryPath) throws IOException {
        ObjectMapper mapper = new ObjectMapper();
        @SuppressWarnings("unchecked")
        List<Map<String, Object>> registry = (List<Map<String, Object>>) mapper.readValue(new File(registryPath), List.class);
        fallbackMap.clear();
        for (Map<String, Object> entry : registry) {
            String selectorKey = (String) entry.get("selectorKey");
            if (selectorKey == null) continue;
            @SuppressWarnings("unchecked")
            List<String> fallbacks = (List<String>) entry.getOrDefault("selectorFallbacks", new ArrayList<>());
            fallbackMap.put(selectorKey, fallbacks);
        }
        loaded = true;
        log.info(String.format("SelectorRegistry loaded: %d entries from %s", fallbackMap.size(), registryPath));
    }

    public static List<String> getFallbacks(String selectorKey) {
        if (!loaded) {
            log.warning("SelectorRegistry.load() was never called — no fallbacks available");
        }
        return fallbackMap.getOrDefault(selectorKey, Collections.emptyList());
    }
}
